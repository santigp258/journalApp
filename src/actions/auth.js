import { types } from "../types/types";


//return login action
export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
