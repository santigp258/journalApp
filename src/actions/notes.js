import Swal from "sweetalert2";

import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
  //async homework
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
    dispatch(startLoadingNotes(uid));
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

//save note in FireStore
export const startSaveNotes = (note) => {
  return async (dispatch, getState) => {
    //user uid
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved!", note.title, "success");
  };
};

//refresh one note
export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: {
    id,
    note: {
      id,
      ...note, //no delete key
    },
  },
});

//upload file

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Please Wait",
      willOpen: () => {
        Swal.showLoading();
      },
    });
    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNotes(activeNote));

    Swal.close();
  };
};

//delete from firestore
export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    await db.doc(`/${uid}/journal/notes/${id}`).delete();
    dispatch(deleteNote(id));
  };
};

//delete note
export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const actionCleaning = () => ({
  type: types.notesActionCleaning,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
