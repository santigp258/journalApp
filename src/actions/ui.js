import { types } from "../types/types";

//using implicit return () => ({})
export const setError = (err) => ({
  type: types.uiSetError,
  payload: err,
});

export const removeError = () => ({
  type: types.uiRemoveError,
});

export const startLoading = () => ({
  type: types.uiStartLoading,
});
export const finishLoading = () => ({
  type: types.uiFinishLoading,
});
