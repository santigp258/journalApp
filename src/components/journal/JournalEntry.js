import React from "react";

export const JournalEntry = () => {
  return (
    <div className="journal__entry mt-1">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: "cover",
          backgroundImage:
            "url(https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/1200px-Image_created_with_a_mobile_phone.png)",
        }}
      ></div>

      <div className="journal__entry-body">
          <p className="journal__entry-title">
              A new day
          </p>
          <div className="journal__entry-content">
              Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum veniam repellendus velit eaque tempora et illo doloremque nostrum esse eius.
          </div>
      </div>
      <div className="journal__entry-date-box">
          <span>Monday</span>
            <h4>28</h4>          
      </div>
    </div>
  );
};
