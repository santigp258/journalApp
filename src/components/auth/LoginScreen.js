import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import { Link } from "react-router-dom";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";

import validator from "validator";
import { removeError, setError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {
  //dispatch for value. React-redux hook
  const dispatch = useDispatch();

  //delete other errors
  useEffect(() => {
    dispatch(removeError("No email"));
  }, [dispatch]);

  //hook useForn
  const [formValues, handleInputChange] = useForm({
    email: "santigp258@gmail.com",
    password: "1234",
  });

  //content ui state
  const { msgError, loading } = useSelector((state) => state.ui);
  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();

    //send action
    if (isFormValid()) {
      dispatch(startLoginEmailPassword(email, password));
    } else {
      console.log("no valid");
    }
  };

  //login with Google
  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  const isFormValid = () => {
    if (!validator.isEmail(email)) {
      dispatch(setError("Is not email. Try valid email"));
    } else if (password.trim().length < 5) {
      dispatch(
        setError(
          "Password should be at least 6 characters and match each other"
        )
      );
      return false;
    } else {
      return true;
    }
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>
      <form onSubmit={handleLogin}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>
        <div className="auth__social-networks">
          <p>Login with Social Network</p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register" className="link">
          Create a new account
        </Link>
      </form>
    </>
  );
};
