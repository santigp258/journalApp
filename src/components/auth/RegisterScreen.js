import React, { useLayoutEffect } from "react";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { useDispatch, useSelector } from "react-redux";
import { removeError, setError } from "../../actions/ui";
import { startRegisterWithEmailPassword } from "../../actions/auth";

export const RegisterScreen = () => {
  //errors
  const dispatch = useDispatch(setError);

  //delete other errors
  useLayoutEffect(() => {
    dispatch(removeError("No email"));
  }, []);

  //content ui state
  const { msgError } = useSelector((state) => state.ui);

  const [formValue, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    password2: "",
  });

  const { name, email, password, password2 } = formValue;

  //register onsubmit
  const handleRegister = (e) => {
    e.preventDefault();
    if (isFormValid()) {
      console.log("Hi!");
      dispatch(startRegisterWithEmailPassword(email, password, name));
    }
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError("Name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("No email"));
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          "Password should be at least 6 characters and match each other"
        )
      );
      return false;
    } else {
      return true;
    }
  };
  return (
    <>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="password"
          onChange={handleInputChange}
          placeholder="Password"
          name="password"
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          Register
        </button>

        <Link to="/auth/login" className="link">
          Already Registered?
        </Link>
      </form>
    </>
  );
};
