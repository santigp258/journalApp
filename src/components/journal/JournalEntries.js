import React from "react";
import { useSelector } from "react-redux";
import { JournalEntry } from "./JournalEntry";

export const JournalEntries = () => {
  const { notes: entries } = useSelector((state) => state.notes);
  return (
    <div className="journal__entries">
      {entries &&
        entries.map((entry) => <JournalEntry key={entry.id} {...entry} />)}
    </div>
  );
};
