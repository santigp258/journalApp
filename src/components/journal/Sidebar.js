import React from "react";
import { JournalEntries } from "./JournalEntries";

export const Sidebar = () => {
  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">
        <h3 className="mt-5">
          <i className="far fa-moon"></i>
          <span>Santiago</span>
        </h3>

        <button className="btn">Logout</button>
      </div>

      <div className="journal__new-entry">
        <i className="fa fa-calendar-plus fa-5x"></i>
        <p>New Entry</p>
      </div>

      <JournalEntries/>
    </aside>
  );
};
