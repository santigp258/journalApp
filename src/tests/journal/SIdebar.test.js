import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router";
import { Sidebar } from "../../components/journal/Sidebar";
import { startLogout } from "../../actions/auth";
import { startNewNote } from "../../actions/notes";
const middlewares = [thunk];
jest.mock("../../actions/auth", () => ({
  startLogout: jest.fn(),
}));

jest.mock("../../actions/notes", () => ({
  startNewNote: jest.fn(),
}));
//create store
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: { loading: false, msgError: null },
  notes: {
    active: { id: "abc" },
    notes: [],
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <Sidebar />
    </MemoryRouter>
  </Provider>
);

describe("pruebas en <Sidebar/>", () => {
  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de llamar el logout", () => {
    wrapper.find("button").prop("onClick")();

    expect(startLogout).toHaveBeenCalledTimes(1);
  });

  test("debe de llamar startNewNote", () => {
    wrapper.find(".journal__new-entry").simulate("click");
    wrapper.find(".journal__new-entry").simulate("click");

    expect(startNewNote).toHaveBeenCalledTimes(2);
  });
});
