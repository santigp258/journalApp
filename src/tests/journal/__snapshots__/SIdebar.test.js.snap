// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <Sidebar/> debe mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span
        className="journal__displayname"
      />
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x pointer"
    />
    <p>
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
